### PR TITLE
8298321: 2 File Leak defect groups in 2 files

### PR DESCRIPTION
**Pull request by calvinccheung**
Please review this simple change for fixing two file leak warnings: 
1) DirectivesParser::parse_from_file_inner() in open/src/hotspot/share/compiler/directivesParser.cpp 
2) mmap_attach_shared() in open/src/hotspot/os/posix/perfMemory_posix.cpp

Please refer to the [comment](https://bugs.openjdk.org/browse/JDK-8298321?focusedCommentId=14549361&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14549361) in the bug report for details.

Passed tiers 1 and 2 testing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298321](https://bugs.openjdk.org/browse/JDK-8298321): 2 File Leak defect groups in 2 files

### Reviewers
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to [754a8870](https://git.openjdk.org/jdk/pull/11909/files/754a88706e22a37eb8e88f1af73493dcbff77979)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to [56980b79](https://git.openjdk.org/jdk/pull/11909/files/56980b7929263f18eb5600be3441ea9f64212889)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11909/head:pull/11909` \
`$ git checkout pull/11909`

Update a local copy of the PR: \
`$ git checkout pull/11909` \
`$ git pull https://git.openjdk.org/jdk pull/11909/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11909`

View PR using the GUI difftool: \
`$ git pr show -t 11909`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11909.diff">https://git.openjdk.org/jdk/pull/11909.diff</a>

</details>

### CODE REVIEW

**Review by iwanowww — APPROVED** [review]
Looks good.

File: src/hotspot/os/posix/perfMemory_posix.cpp [hunk 1]
```diff
@@ -1205,9 +1205,10 @@ static void mmap_attach_shared(int vmid, char** addr, size_t* sizep, TRAPS) {
   FREE_C_HEAP_ARRAY(char, dirname);
   FREE_C_HEAP_ARRAY(char, filename);
 
-  if (fd == OS_ERR || HAS_PENDING_EXCEPTION) {
+  if (fd == OS_ERR) {
     return;
   }
+  assert(!HAS_PENDING_EXCEPTION, "must be");
```
Comment (iklam) [inline src/hotspot/os/posix/perfMemory_posix.cpp:1211]: I think this is better, so you don't need to read the rather convoluted code in open_sharedmem_file() to understand what the API contract is.
```
if (HAS_PENDING_EXCEPTION) {
  assert(fd == OS_ERR, "open_sharedmem_file always returns OS_ERR on exceptions");
}
if (fd == OS_ERR) {
  return;
}
```

**Review by iklam — APPROVED** [review]

File: src/hotspot/os/posix/perfMemory_posix.cpp [hunk 2]
```diff
@@ -1205,7 +1205,10 @@ static void mmap_attach_shared(int vmid, char** addr, size_t* sizep, TRAPS) {
   FREE_C_HEAP_ARRAY(char, dirname);
   FREE_C_HEAP_ARRAY(char, filename);
 
-  if (fd == OS_ERR || HAS_PENDING_EXCEPTION) {
+  if (HAS_PENDING_EXCEPTION) {
+    assert(fd == OS_ERR, "open_sharedmem_file always return OS_ERR on exceptions");
```
Comment (dholmes-ora) [inline src/hotspot/os/posix/perfMemory_posix.cpp:1209]: Does this really appease the static analysis tool?

File: src/hotspot/share/compiler/directivesParser.cpp [hunk 3]
```diff
@@ -102,6 +102,7 @@ bool DirectivesParser::parse_from_file_inner(const char* filename, outputStream*
         ::close(file_handle);
         return parse_string(buffer, stream) > 0;
       }
+      ::close(file_handle);
```
Comment (dholmes-ora) [inline src/hotspot/share/compiler/directivesParser.cpp:105]: You could just move the existing `::close` to before the `if` statement ie close it immediately after the read.

File: src/hotspot/os/posix/perfMemory_posix.cpp (hunk 2 above)
Comment (calvinccheung) [reply src/hotspot/os/posix/perfMemory_posix.cpp:1209]: The static analysis tool should be fine about the above change. As I mentioned in the bug report, the warning is regarding the fd is not closed under the conditions HAS_PENDING_EXCEPTION and fd != OS_ERR.
I will run rerun the test to be sure.

File: src/hotspot/share/compiler/directivesParser.cpp (hunk 3 above)
Comment (calvinccheung) [reply src/hotspot/share/compiler/directivesParser.cpp]: I agree and have pushed another commit with your suggestion.

**Review by dholmes-ora — APPROVED** [review]
Thanks for the update. Pre-approval pending that final check as I'm away for a few days.

File: src/hotspot/os/posix/perfMemory_posix.cpp (hunk 2 above)
Comment (calvinccheung) [reply src/hotspot/os/posix/perfMemory_posix.cpp:1209]: I re-ran tiers 1 and 2 tests using mach5 which tests debug build.
The run with the static analysis tool uses product build.
Both passed.

**Comment by calvinccheung** [comment]
Thanks @iwanowww, @iklam, @dholmes-ora for the review.

/integrate

**Pull request closed without merging**